lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System.Collections.Generic;
+
+namespace Org.Apache.REEF.Common.metrics.Api
+{
+    /// <summary>
+    /// Abstract immutable metric class. All metrics put in the record by the MetricsSource are 
+    /// kept as <see cref="IEnumerable{T}" />.
+    /// </summary>
+    public abstract class AbstractMetric : IMetricsInfo
+    {
+        private readonly IMetricsInfo _info;
+
+        private readonly long _value;
+
+        /// <summary>
+        /// Protected constructor of abstract metric.
+        /// </summary>
+        /// <param name="info">Meta-data for the metric</param>
+        /// <param name="value">The value of the metric. It is assumed that user 
+        /// or application has converted float/double values to long via some logic.</param>
```

Comment:
So, later on in places like IMetricsRecord and elsewhere where we want to store a collection of different metric types (double, float, int, long)  it's much easier to do that as a single collection, although this can be circumvented by making AbstractMetrics derive from another interface/class with no generic types. Moreover, in case we put it as a generic type, we leave the burden of writing the codecs, string representations and other converters of different types to users. In reality, in all cases we will end up using two kinds of types only - long and double. So I do not want to over complicate this.